### PR TITLE
Feat/ecs/engine

### PR DESCRIPTION
**Pull request by aurelenc**
### Feature Ecs' Engine

This pull request add a playable demo of the R-Type.
It also add the following features:

**Worlds**

- [x] add system to Worlds
- [x] add events to Worlds
- [x] Switch between differents worlds
- [x] Set the current world as waiting
- [x] Set the current world as ready to switch

**Engine**

The Engine of the demo mode add all systems and associated components listed below.
It is also able to run them all.

**Systems**

Currently able to add a system to the world & run it.
It also able to display drawable entities on a SFML window and udpate movement and events from user's input.
- [x] Movement system
- [x] Draw system
- [x] HandleSFMLEvents
- [x] HandleSFMLMovements
- [x] ManageClientEvents
- [x] Manage component values using systems

**Components**
The following components are now implemented : 
- [x] Controllable
- [x] Direction
- [x] Drawable
- [x] Size
- [x] Velocity

### Working progress

These following features will be added as soon as possible to the demo

- [ ] Player shoot
- [ ] Damaged enemies
- [ ] Destroy enemies
- [ ] Hibox management

### CODE REVIEW

**Review by Cavonstavant — CHANGES_REQUESTED** [review]
pls add doc

**Comment by aurelenc** [comment]
> pls add doc

no time, will do later bcoz its urgent

**Review by Cavonstavant — APPROVED** [review]
ok

**Review by Bootoyka — APPROVED** [review]
Look good to me

File: src/ecs/Engine.hpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            initWorld.registry.addComponent<ecs::component::Position>(enemy, {500, 500});
+            initWorld.registry.addComponent<ecs::component::Size>(enemy, {5, 5});
+            initWorld.registry.addComponent<ecs::component::Drawable>(
+                enemy, {"src/demo/assets/textures/players.gif", {1, 18, 32, 16}});
+            initWorld.registry.addComponent<ecs::component::EnemyAI>(enemy, {});
+
+            initWorld.addSystem(ecs::systems::handleSFMLEvents);
+            initWorld.addSystem(ecs::systems::handleSFMLMovements);
+            initWorld.addSystem(ecs::systems::manageClientEvents);
+            // initWorld.addSystem(ecs::systems::positionLogger);
```
Comment (Cavonstavant) [inline src/ecs/Engine.hpp:78]: We should consider adding a compiler definitition or some sort of switch to activate when we need to.
How about it @Bootoyka @aurelenc  ?

File: src/ecs/Engine.hpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            if (_waitingWorld)
+                _waitingWorld.release();
+            _waitingWorld = std::make_unique<ecs::World>(world);
+        }
+
+        /**
+         * Set the world given as parameter ready to switch for another one
+         */
+        void setWorldSwitchReady() { _worldSwitchReady = true; }
```
Comment (Cavonstavant) [inline src/ecs/Engine.hpp:105]: consider using an `inline` func
```suggestion
        inline void setWorldSwitchReady() { _worldSwitchReady = true; }
```

File: src/ecs/World.hpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            }
+        }
+
+        /**
+         * USed to add a new system into the world
+         * @param system The system you want to add in the world in a constant manner
+         */
+        void addSystem(const std::function<void(World &)> &system) { _systems.push_back(system); }
+
+        sf::RenderWindow &getWindow() { return *_window; }
```
Comment (Cavonstavant) [inline src/ecs/World.hpp:53]:
```suggestion
        inline sf::RenderWindow &getWindow() { return *_window; }
```

File: src/ecs/World.hpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+        /**
+         * Used to add a new event into the world
+         * @param event The event you want to push in the world
+         */
+        void pushEvent(const ecs::Event event) { _events.push(event); }
```
Comment (Cavonstavant) [inline src/ecs/World.hpp:59]:
```suggestion
        inline void pushEvent(const ecs::Event event) { _events.push(event); }
```

File: src/ecs/systems/ManageClientEvents.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+                    else if (world.getEvent() == ecs::Event::EventType::MoveRight)
+                        dir.value().x += 1;
+                    if (dir.value().x > 0)
+                        dir.value().x = 1;
+                    else if (dir.value().x < 0)
+                        dir.value().x = -1;
+                    if (dir.value().y > 0)
+                        dir.value().y = 1;
+                    else if (dir.value().y < 0)
+                        dir.value().y = -1;
```
Comment (Cavonstavant) [inline src/ecs/systems/ManageClientEvents.hpp:53]: **Warning**, these checks only for screen limits on on side (left)

File: src/ecs/systems/Movement.hpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        for (size_t i = 0; i < positions.size() && i < velocities.size(); ++i) {
+            auto &pos = positions[i];
+            auto const &vel = velocities[i];
+            auto &dir = directions[i];
+            if (pos && vel) {
+                if (clock.getElapsedTime().asMilliseconds() > 10) {
+                    pos.value().x += vel.value().x * (dir.value().x);
+                    pos.value().y += vel.value().y * dir.value().y;
+                    clock.restart();
+                    dir.value().x = 0;
```
Comment (Cavonstavant) [inline src/ecs/systems/Movement.hpp:43]: Maybe we should add some getters and setters, idk

**Review by Cavonstavant — APPROVED** [review]
Everything s looks alright, just some incomprehensions on my side -_-

File: src/ecs/Engine.hpp (hunk 1 above)
Comment (Bootoyka) [reply src/ecs/Engine.hpp:78]: Are u talking about systems ? if yes understand that they are continualy running in the current world

File: src/ecs/systems/ManageClientEvents.hpp (hunk 5 above)
Comment (Bootoyka) [reply src/ecs/systems/ManageClientEvents.hpp:53]: @aurelenc

Comment (aurelenc) [reply src/ecs/systems/ManageClientEvents.hpp:53]: That's not about the screen, read the variable names

**Pull request merged**